RevenueCatUI/CustomerCenter/Data/LoadPromotionalOfferUseCase.swift

```diff
@@ -43,47 +43,103 @@ class LoadPromotionalOfferUseCase: LoadPromotionalOfferUseCaseType {
-                  let subscribedProduct = await self.purchasesProvider.products([productIdentifier]).first else {
-                Logger.warning(Strings.could_not_offer_for_any_active_subscriptions)
-                return .failure(CustomerCenterError.couldNotFindSubscriptionInformation)
-            }
-
-            let exactMatch = subscribedProduct.discounts.first { discount in
-                discount.offerIdentifier == promoOfferDetails.iosOfferId
-            }
-
-            let discount: StoreProductDiscount?
-            if let exactMatch = exactMatch {
-                discount = exactMatch
-            } else {
-                discount = subscribedProduct.discounts.first { discount in
-                    guard let offerIdentifier = discount.offerIdentifier else {
-                        return false
-                    }
-                    return offerIdentifier.hasSuffix("_\(promoOfferDetails.iosOfferId)")
-                }
-            }
-
-            guard let discount = discount else {
-                let message =
-                Strings.could_not_offer_for_active_subscriptions(promoOfferDetails.iosOfferId, productIdentifier)
-                Logger.debug(message)
-                return .failure(CustomerCenterError.couldNotFindSubscriptionInformation)
-            }
-
-            let promotionalOffer = try await self.purchasesProvider.promotionalOffer(forProductDiscount: discount,
-                                                                                     product: subscribedProduct)
-            let promotionalOfferData = PromotionalOfferData(promotionalOffer: promotionalOffer,
-                                                            product: subscribedProduct,
-                                                            promoOfferDetails: promoOfferDetails)
-            return .success(promotionalOfferData)
+            let (productIdentifier, subscribedProduct) = try await getActiveSubscription(customerInfo)
+            let discount = try findDiscount(for: subscribedProduct,
+                                            productIdentifier: productIdentifier,
+                                            promoOfferDetails: promoOfferDetails)
+
+            let promotionalOffer = try await self.purchasesProvider.promotionalOffer(
+                forProductDiscount: discount,
+                product: subscribedProduct
+            )
+            return .success(PromotionalOfferData(
+                promotionalOffer: promotionalOffer,
+                product: subscribedProduct,
+                promoOfferDetails: promoOfferDetails
+            ))
         } catch {
             Logger.warning(Strings.error_fetching_promotional_offer(error))
             return .failure(CustomerCenterError.couldNotFindOfferForActiveProducts)
         }
     }
 
+    private func getActiveSubscription(_ customerInfo: CustomerInfo) async throws -> (String, StoreProduct) {
```

Comment:
I don't feel particularly strong about these things so FFTI but in our style guide we suggest moving [private functions to a private extension](https://github.com/RevenueCat/purchases-ios/blob/main/Contributing/SwiftStyleGuide.swift#L119)